x/ecocredit/core/genesis.go

```diff
@@ -0,0 +1,267 @@
+	defer bItr.Close()
+
+	// create index batchID => precision for faster lookup
+	for bItr.Next() {
+		batch, err := bItr.Value()
+		if err != nil {
+			return err
+		}
+
+		if _, exists := batchIdToPrecision[batch.Id]; exists {
+			continue
+		}
+
+		class, err := ss.ClassInfoTable().Get(ormCtx, projectIdToClassId[batch.ProjectId])
+		if err != nil {
+			return err
+		}
+
+		if class.Id == projectIdToClassId[batch.ProjectId] {
+			batchIdToPrecision[batch.Id] = abbrevToPrecision[class.CreditType]
+		}
+	}
+
+	batchIdToCalSupply := make(map[uint64]math.Dec) // map of batchID to calculated supply
+	batchIdToSupply := make(map[uint64]math.Dec)    // map of batchID to actual supply
+	bsItr, err := ss.BatchSupplyTable().List(ormCtx, api.BatchSupplyPrimaryKey{})
+	if err != nil {
+		return err
+	}
+	defer bsItr.Close()
+
+	// calculate total supply for each credit batch (tradable + retired supply)
+	for bsItr.Next() {
+		batchSupply, err := bsItr.Value()
+		if err != nil {
+			return err
+		}
+
+		tSupply := math.NewDecFromInt64(0)
+		rSupply := math.NewDecFromInt64(0)
+		if batchSupply.TradableAmount != "" {
+			tSupply, err = math.NewNonNegativeFixedDecFromString(batchSupply.TradableAmount, batchIdToPrecision[batchSupply.BatchId])
+			if err != nil {
+				return err
+			}
+		}
+		if batchSupply.RetiredAmount != "" {
+			rSupply, err = math.NewNonNegativeFixedDecFromString(batchSupply.RetiredAmount, batchIdToPrecision[batchSupply.BatchId])
+			if err != nil {
+				return err
+			}
+		}
+
+		total, err := math.SafeAddBalance(tSupply, rSupply)
+		if err != nil {
+			return err
+		}
+
+		batchIdToSupply[batchSupply.BatchId] = total
+	}
+
+	// calculate credit batch supply from genesis tradable and retired balances
+	if err := calculateSupply(ormCtx, batchIdToPrecision, ss, batchIdToCalSupply); err != nil {
+		return err
+	}
+
+	// verify calculated total amount of each credit batch matches the total supply
+	if err := validateSupply(batchIdToCalSupply, batchIdToSupply); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func calculateSupply(ctx context.Context, decimalPlaces map[uint64]uint32, ss api.StateStore, calSupply map[uint64]math.Dec) error {
+	bbItr, err := ss.BatchBalanceTable().List(ctx, api.BatchBalancePrimaryKey{})
+	if err != nil {
+		return err
+	}
+	defer bbItr.Close()
+
+	for bbItr.Next() {
+		tBalance := math.NewDecFromInt64(0)
+		rBalance := math.NewDecFromInt64(0)
+
+		balance, err := bbItr.Value()
+		if err != nil {
+			return err
+		}
+
+		if _, ok := decimalPlaces[balance.BatchId]; !ok {
+			return sdkerrors.ErrInvalidType.Wrapf("credit type not exist for %d batch", balance.BatchId)
+		}
+
+		if balance.Tradable != "" {
+			tBalance, err = math.NewNonNegativeFixedDecFromString(balance.Tradable, decimalPlaces[balance.BatchId])
+			if err != nil {
+				return err
+			}
+		}
+
+		if balance.Retired != "" {
+			rBalance, err = math.NewNonNegativeFixedDecFromString(balance.Retired, decimalPlaces[balance.BatchId])
+			if err != nil {
```

Comment:
we should also add one for `Escrowed` amounts. these are differenciated on the batch balance level, but are part of the the `TradableSupply` amount. so:

supply tradable + supply retired = balances tradable + balances retired + balances escrowed